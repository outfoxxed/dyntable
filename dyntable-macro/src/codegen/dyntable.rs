use proc_macro2::TokenStream;
use quote::format_ident;
use syn::{Signature, Token};

use crate::parse::{DynTraitInfo, VTableEntry};

pub fn impl_dyntable(dyntrait: &DynTraitInfo) -> TokenStream {
	let proxy_trait = format_ident!("__DynTable_{}", dyntrait.dyntrait.ident);
	let ident = &dyntrait.dyntrait.ident;
	let vtable_ident = &dyntrait.vtable.name;
	let (_, ty_generics, where_clause) = dyntrait.generics.split_for_impl();
	let impl_generics = dyntrait.generics.params.clone().into_iter()
    .collect::<Vec<_>>();
	let where_entries = where_clause
    .into_iter()
    .flat_map(|clause| &clause.predicates);

	let vtable_entries = dyntrait.entries.iter()
    .map(|entry| match entry {
			VTableEntry::Subtable(subtable) => {
				let entry_name = &subtable.ident;
				quote::quote! {
					#entry_name: <__DynTarget as ::dyntable::DynTable<
						<(dyn #ident #ty_generics + 'static) as ::dyntable::VTableRepr>::VTable,
					>>::VTABLE
				}
			},
			VTableEntry::Method(method) => {
				let Signature {
					unsafety,
					abi,
					fn_token,
					ident,
					inputs,
					variadic,
					output,
					..
				} = &method.sig;

				let inputs = inputs.iter()
					.map(|_| <Token![_]>::default());

				let output = match output {
					syn::ReturnType::Default => None,
					syn::ReturnType::Type(..) => Some(<Token![_]>::default()),
				}.into_iter();

				quote::quote! {
					#ident: unsafe {
						::core::intrinsics::transmute(
							__DynTarget::#ident as
								#unsafety #abi #fn_token (
									#(#inputs,)* #variadic
								) #( -> #output)*
						)
					}
				}
			},
		});

	quote::quote! {
		#[allow(non_camel_case_types)]
		unsafe trait #proxy_trait<'v, V: 'v + ::dyntable::VTable> {
			const VTABLE: V;
			const STATIC_VTABLE: &'v V;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget,
		> ::dyntable::__private::DynTable2<'__dyn_vtable, #vtable_ident #ty_generics>
		for ::dyntable::__private::DynImplTarget<__DynTarget, #vtable_ident #ty_generics>
		where
			#(#where_entries,)*
			__DynTarget: #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>,
		{
			const VTABLE: #vtable_ident #ty_generics = __DynTarget::VTABLE;
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics = __DynTarget::STATIC_VTABLE;
		}

		unsafe impl<
			'__dyn_vtable,
			#(#impl_generics + '__dyn_vtable,)*
			__DynTarget: #ident #ty_generics,
		> #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>
		for __DynTarget
		#where_clause {
			const STATIC_VTABLE: &'__dyn_vtable #vtable_ident #ty_generics =
				&<Self as #proxy_trait<'__dyn_vtable, #vtable_ident #ty_generics>>::VTABLE;
			const VTABLE: #vtable_ident #ty_generics = #vtable_ident {
				#(#vtable_entries,)*
			};
		}
	}
}
