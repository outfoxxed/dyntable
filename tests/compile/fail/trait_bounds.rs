//! This test ensures that a #[dyntable] trait cannot have implicit
//! bound inheritance (will hide implemented methods from the user),
//! or have bounds on non #[dyntable] traits.

// FIXME: extremely confusing or misleading error messages

use dyntable::dyntable;

fn main() {}

trait NonDyntableTrait {}

/// Inherit a non dyn trait without a dyn entry in the where clause
#[dyntable]
trait NonDynBound: NonDyntableTrait {}

/// Inherit a non dyn trait with a dyn entry in the where clause
#[dyntable]
trait NonDynBoundWithClause: NonDyntableTrait
where
	dyn NonDyntableTrait:,
{}

// --

#[dyntable]
trait DyntableTrait {}

/// Inherit a dyn trait without adding a dyn entry in the where clause
#[dyntable]
trait DynBoundWithoutClause: DyntableTrait {}

// --

#[dyntable]
trait RequiredTrait {}

#[dyntable]
trait ImplementedTrait: RequiredTrait
where
	dyn RequiredTrait:,
{}

/// Inherit a dyn trait without explicitly specifying its bounds in
/// its dyn entry in the where clause
#[dyntable]
trait MissingExplicitBoundInheritance: ImplementedTrait
where
	dyn ImplementedTrait:,
{}
