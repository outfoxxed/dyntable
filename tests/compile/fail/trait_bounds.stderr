error[E0277]: the trait bound `Dyn<__DynRepr>: NonDyntableTrait` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:14:1
   |
14 | #[dyntable]
   | ^^^^^^^^^^^ the trait `NonDyntableTrait` is not implemented for `Dyn<__DynRepr>`
   |
note: required by a bound in `NonDynBound`
  --> tests/compile/fail/trait_bounds.rs:15:20
   |
15 | trait NonDynBound: NonDyntableTrait {}
   |                    ^^^^^^^^^^^^^^^^ required by this bound in `NonDynBound`
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `(dyn NonDyntableTrait + 'static): VTableRepr` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:18:1
   |
18 | #[dyntable]
   | ^^^^^^^^^^^ the trait `VTableRepr` is not implemented for `(dyn NonDyntableTrait + 'static)`
   |
   = help: the following other types implement trait `VTableRepr`:
             (dyn DynBoundWithoutClause + 'static)
             (dyn DynBoundWithoutClause + Send + 'static)
             (dyn DynBoundWithoutClause + Send + Sync + 'static)
             (dyn DynBoundWithoutClause + Sync + 'static)
             (dyn DyntableTrait + 'static)
             (dyn DyntableTrait + Send + 'static)
             (dyn DyntableTrait + Send + Sync + 'static)
             (dyn DyntableTrait + Sync + 'static)
           and $N others
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `Dyn<__DynRepr>: NonDyntableTrait` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:18:1
   |
18 | #[dyntable]
   | ^^^^^^^^^^^ the trait `NonDyntableTrait` is not implemented for `Dyn<__DynRepr>`
   |
note: required by a bound in `NonDynBoundWithClause`
  --> tests/compile/fail/trait_bounds.rs:19:30
   |
19 | trait NonDynBoundWithClause: NonDyntableTrait
   |                              ^^^^^^^^^^^^^^^^ required by this bound in `NonDynBoundWithClause`
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: the trait bound `__DynSubTables: SubTable<NonDynBoundWithClauseVTable>` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:18:1
   |
18 | #[dyntable]
   | ^^^^^^^^^^^ the trait `SubTable<NonDynBoundWithClauseVTable>` is not implemented for `__DynSubTables`
   |
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting type parameter `__DynSubTables`
   |
18 | #[dyntable], __DynSubTables: dyntable::SubTable<NonDynBoundWithClauseVTable>
   |            +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `__DynSubTables: SubTable<DyntableTraitVTable>` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:30:1
   |
30 | #[dyntable]
   | ^^^^^^^^^^^ the trait `SubTable<DyntableTraitVTable>` is not implemented for `__DynSubTables`
   |
note: required for `Dyn<__DynRepr>` to implement `DyntableTrait`
  --> tests/compile/fail/trait_bounds.rs:26:1
   |
26 | #[dyntable]
   | ^^^^^^^^^^^
note: required by a bound in `DynBoundWithoutClause`
  --> tests/compile/fail/trait_bounds.rs:31:30
   |
31 | trait DynBoundWithoutClause: DyntableTrait {}
   |                              ^^^^^^^^^^^^^ required by this bound in `DynBoundWithoutClause`
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting type parameter `__DynSubTables`
   |
30 | #[dyntable], __DynSubTables: dyntable::SubTable<DyntableTraitVTable>
   |            +++++++++++++++++++++++++++++++++++++++++++++++++++++++++

error[E0277]: the trait bound `__DynSubTables: SubTable<RequiredTraitVTable>` is not satisfied
  --> tests/compile/fail/trait_bounds.rs:46:1
   |
46 | #[dyntable]
   | ^^^^^^^^^^^ the trait `SubTable<RequiredTraitVTable>` is not implemented for `__DynSubTables`
   |
note: required for `Dyn<__DynRepr>` to implement `ImplementedTrait`
  --> tests/compile/fail/trait_bounds.rs:38:1
   |
38 | #[dyntable]
   | ^^^^^^^^^^^
note: required by a bound in `MissingExplicitBoundInheritance`
  --> tests/compile/fail/trait_bounds.rs:47:40
   |
47 | trait MissingExplicitBoundInheritance: ImplementedTrait
   |                                        ^^^^^^^^^^^^^^^^ required by this bound in `MissingExplicitBoundInheritance`
   = note: this error originates in the attribute macro `dyntable` (in Nightly builds, run with -Z macro-backtrace for more info)
help: consider further restricting type parameter `__DynSubTables`
   |
46 | #[dyntable], __DynSubTables: dyntable::SubTable<RequiredTraitVTable>
   |            +++++++++++++++++++++++++++++++++++++++++++++++++++++++++
